### PR TITLE
feat: implement mirror and oci settings

### PR DESCRIPTION
**Pull request by wolfv**
This also updates the `toml` crate because I ran `cargo update`. Apparently `Document` is deprecated in favor of `DocumentMut`.

### CODE REVIEW

File: docs/advanced/global_configuration.md [hunk 1]
```diff
@@ -29,4 +32,61 @@ change_ps1 = true
...
+
+```toml
+[mirrors]
+"https://conda.anaconda.org": [
+    "https://prefix.dev/"
+]
+```
+
+This will forward all request to channels on anaconda.org to prefix.dev.
+Channels that are not currently mirrored on prefix.dev will fail.
```
Comment (pavelzw) [inline docs/advanced/global_configuration.md:76]:
```suggestion
Channels that are not currently mirrored on prefix.dev will fail in the above example.
```

**Comment by pavelzw** [comment]
Closes #172

File: docs/advanced/global_configuration.md [hunk 2]
```diff
@@ -29,4 +32,61 @@ change_ps1 = true
...
+]
+
+# redirect all requests for bioconda to one of the three listed mirrors
+# Note: for repodata we try the first mirror first.
+"https://conda.anaconda.org/bioconda": [
+    "https://conda.anaconda.org/bioconda",
+    # OCI registries are also supported
+    "oci://ghcr.io/channel-mirrors/bioconda",
+    "https://prefix.dev/bioconda",
+]
```
Comment (ruben-arts) [inline docs/advanced/global_configuration.md:49]:
```suggestion
"https://conda.anaconda.org/conda-forge"= [
    "https://prefix.dev/conda-forge"
]

# redirect all requests for bioconda to one of the three listed mirrors
# Note: for repodata we try the first mirror first.
"https://conda.anaconda.org/bioconda"= [
    "https://conda.anaconda.org/bioconda",
    # OCI registries are also supported
    "oci://ghcr.io/channel-mirrors/bioconda",
    "https://prefix.dev/bioconda",
]
```

Also testing this example gives me:
```
➜ pixi add boltons
ERROR rattler_repodata_gateway::fetch: error=repodata not found
ERROR rattler_repodata_gateway::fetch: error=repodata not found
ERROR rattler_repodata_gateway::fetch: error=repodata not found
ERROR rattler_repodata_gateway::fetch: error=repodata not found
ERROR rattler_repodata_gateway::fetch: error=repodata not found
  × failed to fetch repodata from channels
  ├─▶ repodata not found
  ├─▶ Request error: HTTP status client error (404 Not Found) for url (https://prefix.dev/noarch/repodata.json.zst)
  ╰─▶ HTTP status client error (404 Not Found) for url (https://prefix.dev/noarch/repodata.json.zst)

```

File: Cargo.toml [hunk 3]
```diff
@@ -157,17 +157,19 @@ toml = "0.8.10"
...
-#rattler_conda_types = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_digest = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_lock = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_networking = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_repodata_gateway = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_shell = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_solve = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#rattler_virtual_packages = { git = "https://github.com/mamba-org/rattler", rev = "7e0a130f0603fc2ff204649df9adba78422de2c8" }
-#deno_task_shell = { path = "../deno_task_shell" }#rattler = { path = "../rattler/crates/rattler" }
+rattler = { git = "https://github.com/mamba-org/rattler", rev = "e3784e0bc65b98f3cf3dcb7e2b8924113f3ec463" }
```
Comment (ruben-arts) [inline Cargo.toml:162]: Can this now be a rattler release?

File: docs/advanced/global_configuration.md [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
+   dependent on XDG_CONFIG_HOME)
+2. Global .pixi folder: `~/.pixi/config.toml` (or `$PIXI_HOME/config.toml` if
+   the `PIXI_HOME` environment variable is set)
 3. Project-local .pixi folder: `$PIXI_PROJECT/.pixi/config.toml`
 4. Command line arguments (`--tls-no-verify`, `--change-ps1=false` etc.)
 
-!!! note
-    To find the locations where `pixi` looks for configuration files, run `pixi` with `-v` or `--verbose`.
+!!! note To find the locations where `pixi` looks for configuration files, run
+    `pixi` with `-v` or `--verbose`.
```
Comment (ruben-arts) [inline docs/advanced/global_configuration.md:15]:
```suggestion
!!! note 
    To find the locations where `pixi` looks for configuration files, run
    `pixi` with `-v` or `--verbose`.
```
This broke the note.

File: docs/advanced/global_configuration.md [hunk 5]
```diff
@@ -29,4 +32,61 @@ change_ps1 = true
...
+important to get this file from a trusted source.
+
+You can also specify mirrors for an entire "host", e.g.
+
+```toml
+[mirrors]
+"https://conda.anaconda.org": [
+    "https://prefix.dev/"
+]
+```
```
Comment (ruben-arts) [inline docs/advanced/global_configuration.md:73]:
```suggestion
```toml
[mirrors]
"https://conda.anaconda.org"= [
    "https://prefix.dev/"
]
```
```

File: docs/advanced/global_configuration.md [hunk 6]
```diff
@@ -29,4 +32,61 @@ change_ps1 = true
...
+
+You can also specify mirrors on the OCI registry. There is a public mirror on
+the Github container registry (ghcr.io) that is maintained by the conda-forge
+team. You can use it like this:
+
+```toml
+[mirrors]
+"https://conda.anaconda.org/conda-forge": [
+    "oci://ghcr.io/channel-mirrors/conda-forge"
+]
```
Comment (ruben-arts) [inline docs/advanced/global_configuration.md:88]:
```suggestion
"https://conda.anaconda.org/conda-forge" = [
    "oci://ghcr.io/channel-mirrors/conda-forge"
]
```

Comment (ruben-arts) [reply docs/advanced/global_configuration.md:88]: This also give the same error with the channel repodata, I guess you are missing the name of the channel.

Comment (wolfv) [reply docs/advanced/global_configuration.md]: Fixed by always forcibly adding a trailing `/` - good catch :)

File: docs/advanced/global_configuration.md (hunk 4 above)
Comment (wolfv) [reply docs/advanced/global_configuration.md:15]: fixed too.

**Comment by pavelzw** [comment]
could we also add a `info` or `debug` message that tells you which mirror you are using in the end?

**Comment by ruben-arts** [comment]
Testing the examples in the documentation still gives me:

```
ERROR rattler_repodata_gateway::fetch: error=Request error: HTTP status client error (401 Unauthorized) for url (https://prefix.dev/conda-forge/osx-arm64/repodata.json.zst)
  × failed to fetch repodata from channels
  ├─▶ Request error: HTTP status client error (401 Unauthorized) for url (https://prefix.dev/conda-forge/osx-arm64/repodata.json.zst)
  ╰─▶ HTTP status client error (401 Unauthorized) for url (https://prefix.dev/conda-forge/osx-arm64/repodata.json.zst)
```
Not sure it is because the website doesn't work or because of the mirror.

File: src/utils/reqwest.rs [hunk 7]
```diff
@@ -34,6 +36,42 @@ fn auth_middleware(config: &Config) -> AuthenticationMiddleware {
...
+    }
+
+    for (key, value) in config.mirror_map() {
+        let mut mirrors = Vec::new();
+        for v in value {
+            mirrors.push(Mirror {
+                url: ensure_trailing_slash(v),
+                no_jlap: false,
+                no_bz2: false,
+                no_zstd: false,
```
Comment (pavelzw) [inline src/utils/reqwest.rs:61]: Shouldn't these be dependent on the `config`?

Comment (pavelzw) [reply src/utils/reqwest.rs:61]: Ah nvm this should be dont in #1015

**Pull request merged**